doc/js-perspective-on-c.md

```diff
@@ -0,0 +1,108 @@
+
+With `c` as the context 'a' is uninitialized and can be dangerous. 'b' is invalid syntax and 'c' is an integer initialized to 42.
+
+With `c++` as the context 'a' is treated the same as raw `c`. 'b' is an integer initialized to 42 using constructor syntax. 'c' is an integer initialized to 42 using assignment syntax.
+
+## Passing To and Returning From Functions
+
+Building functions happen for the same reasons in c as they do in JS.  They help to ease the mental burden for "what something does".  The differences in memory allocation drive very different coding paradigms though.
+
+In modern JS we generally pass in what is needed inside the function context, return back the values that result and if an error occurs it is thrown.  That pattern generally holds true for both sync an async functions.
+
+`C` stack-allocation semantics, and the resulting variable scoping, often result in a pattern where both input and output arguments are passed and then the function either returns `void` or `int` (to denote the error code where `0` represents no error) to show if valid execution occurred.  Generally one sees variables allocated for both inputs and return(s) in the calling context and then those variables are passed into the function.  Many times the function call is placed in a conditional to check execution status.
+
+```c
+KZGSettings settings; // this is the output of load_trusted_setup_file()
+FILE *file_handle = fopen(******, "r"); // this is the input of load_trusted_setup_file()
+C_KZG_RET ret = load_trusted_setup_file(&settings, file_handle);
+
+if (fclose(file_handle) != 0) {
+    if (ret == C_KZG_OK) {
+        free_trusted_setup(&(data->settings));
+    }
+}
+
+if (ret != C_KZG_OK) {
+    return 1;
+}
+
+return 0;
+```
+
+There is one more paradigm that is worth explicitly stating.  Note in this example that `fclose` is in the conditional and `load_trusted_setup_file` is not.  `load_trusted_setup_file` is a function where the return value needs to be handled for either valid or invalid execution (ie valid but if fclose fails need to `free_trusted_setup` and for invalid execution the function returns `1`).  `fclose` however only requires that invalid execution requires further work.
+
+There are a few challenges that emerge from the way that `c` functions are structured.  In very large code-bases its very difficult to track allocations and de-allocations because they do not always occur in the same functions.
+
+In `c++` everything from above is true and there are some additional paradigms that should be presented.  Object have higher complexity for initialization/clean-up and bigger sizes.  Those "enhancements" and the `class` paradigm are where most of the complexity in `c++` (relative to `c`) comes in.  The trade of for the complexity is RAII.  Or i suppose RAII emerged because of the complexity and it is a very nice paradigm for large and complex systems.
```

Comment:
```suggestion
In `C++`, everything from above is true, and there are some additional paradigms that should be presented.  Objects have higher complexity for initialization/clean-up and bigger sizes.  Those "enhancements" and the `class` paradigm are where most of the complexity in `C++` (relative to `C`) comes in.  The trade off for the complexity is RAII.  Or I suppose RAII emerged because of the complexity and it is a very nice paradigm for large and complex systems.
```

Reply:
Updated.  Good catches!  I fixed everything but did not add the `,` after true to keep the cadence up in a long section.  Will that be ok?